### PR TITLE
add rule_engine, first match policy coverage

### PR DESCRIPTION
**Pull request by ashishgaurav13**
for #293

### CODE REVIEW

File: lib/ssh_scan/policy.rb [hunk 1]
```diff
@@ -25,5 +26,20 @@ def self.from_string(string)
       opts = YAML.load(string)
       self.new(opts)
     end
+
+    # Choose a policy file based on rules
+    def self.first_match(result)
+      Dir.glob("policies/**/*.yml") do |policy_file|
```
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: This might not work once we release the code as a gem, consider expanding the path based on current location of this file.

File: lib/ssh_scan/policy.rb [hunk 2]
```diff
@@ -25,5 +26,20 @@ def self.from_string(string)
...
+        next if policy_file == @fallback_file
+        file = YAML.load_file(policy_file)
+        next unless file["rules"]
+        puts "Matching: #{policy_file}"
+        if SSHScan::RuleEngine.match(file["rules"], result)
+          puts "Matched: #{policy_file}"
```
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: Same here, RE: logger

File: lib/ssh_scan/policy.rb (hunk 1 above)
Comment (agaurav77) [inline lib/ssh_scan/policy.rb]: Since ssh_scan first chdirs to the root directory, and it is the only script calling this function, it should work, right?

File: lib/ssh_scan/policy.rb [hunk 3]
```diff
@@ -25,5 +26,20 @@ def self.from_string(string)
...
+          policy_file
```
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: do you want "return" before this?

Comment (agaurav77) [inline lib/ssh_scan/policy.rb]: I can do that to make it clearer, but essentially return ... will do the same thing, right?

File: lib/ssh_scan/policy.rb [hunk 4]
```diff
@@ -14,6 +14,7 @@ def initialize(opts = {})
       @references = opts['references'] || []
       @auth_methods = opts['auth_methods'] || []
       @ssh_version = opts['ssh_version'] || false
+      @fallback_file = "policies/mozilla_modern.yml"
```
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: Same here, with the file reference.

File: lib/ssh_scan/rule_engine.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+              answer &= (result[attribute] > value)
+            elsif conditional == "lt"
+              answer &= (result[attribute] < value)
+            elsif conditional == "gte"
+              answer &= (result[attribute] >= value)
+            elsif conditional == "lte"
+              answer &= (result[attribute] <= value)
+            end
+          end
+        end
```
Comment (claudijd) [inline lib/ssh_scan/rule_engine.rb]: This block of nested if's might be a better fit as a case statement or it's own unique object class that can cleanly parse a single rule.

File: spec/ssh_scan/rule_engine_spec.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+require 'rspec'
+require 'ssh_scan/rule_engine'
+
+describe SSHScan::RuleEngine do
+  context "for conditionals" do
+    it "eq, neq should work for any literal" do
+      result = {:ssh_lib => "openssh"}
+      expect(SSHScan::RuleEngine.match({"ssh_lib_eq" => "openssh"}, result)).to be_truthy
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: recommend be_true here.   In ruby, "true" is not the same as "truthy".

Since your match routine is always true/false, I would recommend constraining your unit-tests to look for those specifically.

Reference: for more truthy education (https://gist.github.com/jfarmer/2647362)

Comment (agaurav77) [inline spec/ssh_scan/rule_engine_spec.rb]: Will change, did not know about this one. Thanks!

File: lib/ssh_scan/policy.rb (hunk 3 above)
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: In ruby, you get an implicit return if it's the last operation in the scope.

def example
   foo #will return foo, but implicitly
end

def example
   return foo #will also return foo, but explicitly
end

but if you want an early return, you need to be explicit...

def example
  if foo
    return bar #if foo, then return bar
  end

  foo # if not foo, then default to implicit return of foo
end

File: lib/ssh_scan/policy.rb (hunk 1 above)
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: My understanding would be that it will work when you run from source, but not when you run as a gem.

You can test this by doing the following:

1.) gem build ssh_scan.gemspec # build a gem
2.) gem install ssh_scan-0.0.16.gem # install the gem
3.) go to a different directory and run to exercise this code in IRB and see if the path still works

File: lib/ssh_scan/policy.rb (hunk 3 above)
Comment (agaurav77) [inline lib/ssh_scan/policy.rb]: Didn't know this difference, thanks for pointing out! Changing..

**Comment by agaurav77** [comment]
@claudijd please re review

File: spec/ssh_scan/rule_engine_spec.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+require 'rspec'
+require 'ssh_scan/rule_engine'
+
+describe SSHScan::RuleEngine do
+  context "for conditionals" do
+    it "eq, neq should work for any literal" do
+      result = {:ssh_lib => "openssh"}
+      expect(SSHScan::RuleEngine.match({"ssh_lib_eq" => "openssh"}, result)).to be true
+      expect(SSHScan::RuleEngine.match({"ssh_lib_neq" => "dropbear"}, result)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: I would recommend adding this test as well...

`expect(SSHScan::RuleEngine.match({"ssh_lib_eq" => "dropbear"}, result)).to be false`

File: spec/ssh_scan/rule_engine_spec.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.match({"ssh_version_eq" => 1.99}, result)).to be false
+      expect(SSHScan::RuleEngine.match({"ssh_version_neq" => 1.99}, result)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: I would recommend adding this test as well...

`expect(SSHScan::RuleEngine.match({"ssh_version_eq" => 2.0}, result)).to be true`

File: spec/ssh_scan/rule_engine_spec.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    it "gt, lt, gte, lte should work for numeric literal" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.match({"ssh_version_gt" => 1.99}, result)).to be true
+      expect(SSHScan::RuleEngine.match({"ssh_version_gte" => 2.0}, result)).to be true
+      expect(SSHScan::RuleEngine.match({"ssh_version_lt" => 1.99}, result)).to be false
+      expect(SSHScan::RuleEngine.match({"ssh_version_lte" => 2.0}, result)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Recommend testing below/eql/above examples (to verify boundaries) for each verb, like so...

    # Testing gt boundaries
    expect(SSHScan::RuleEngine.match({"ssh_version_gt" => 1.99}, result)).to be true
    expect(SSHScan::RuleEngine.match({"ssh_version_gt" => 2.0}, result)).to be false
    expect(SSHScan::RuleEngine.match({"ssh_version_gt" => 2.1}, result)).to be false

    # Testing gte boundaries
    expect(SSHScan::RuleEngine.match({"ssh_version_gte" => 1.99}, result)).to be true
    expect(SSHScan::RuleEngine.match({"ssh_version_gte" => 2.0}, result)).to be true
    expect(SSHScan::RuleEngine.match({"ssh_version_gte" => 2.1}, result)).to be false

    # Testing lt boundaries
    expect(SSHScan::RuleEngine.match({"ssh_version_lt" => 1.99}, result)).to be false
    expect(SSHScan::RuleEngine.match({"ssh_version_lt" => 2.0}, result)).to be false
    expect(SSHScan::RuleEngine.match({"ssh_version_lt" => 2.1}, result)).to be true

    # Testing lte boundaries
    expect(SSHScan::RuleEngine.match({"ssh_version_lte" => 1.99}, result)).to be false
    expect(SSHScan::RuleEngine.match({"ssh_version_lte" => 2.0}, result)).to be true
    expect(SSHScan::RuleEngine.match({"ssh_version_lte" => 2.1}, result)).to be true

File: spec/ssh_scan/rule_engine_spec.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    end
+
+    it "should ignore attributes not in results" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.match({"ssh_lib_eq" => "openssh"}, result)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Shouldn't the test expectation here be "false" and not "true"?

Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: My expectation here is that you need to have two elements that match before a match would be "true", but the example case is matching true to two different elements.

**Comment by agaurav77** [comment]
@claudijd 

1. Most of the online boolean expression parsers were either broken or overkill for simple expression parsing, so I created a simple one using an array-based stack.

2. I have added sufficient tests to check if the expression parsing is okay, feel free to play around with the tests.

3. It is very easy to see how the policy file first match would work (like in the updated [wiki](https://github.com/mozilla/ssh_scan/wiki/Policy-Expression-DSL)). Just create a file in the `policies` folder (copy it over from say `mozilla_intermediate.yml`). Add in a few conditions like

    ```yaml
    rules:
      a:
        - ssh_lib_eq
        - openssh
      b:
        - ssh_version_gt
        - 1.99
    ruleset: "!a && ((b))" # parentheses supported
    ...
    ```
    When you run it on an endpoint, it should try to match against this file (and will show it in the logs, stdout/stderr).

    I am going with this notation for `rules` because if I instead have something like `"ssh_lib_eq openssh"`, I have to additionally parse the value. YAML conversion handles it if I do it this way.

I hope this should be sufficient for #293. Also do checkout the checklist in the updated [wiki](https://github.com/mozilla/ssh_scan/wiki/Policy-Expression-DSL)!

File: bin/ssh_scan [hunk 11]
```diff
@@ -127,6 +126,12 @@ scan") do |file|
   opts.on("-P", "--policy [FILE]",
           "Custom policy file (Default: Mozilla Modern)") do |policy|
     options["policy"] = policy
+    unless File.exist?(options["policy"])
+      puts opt_parser.help
+      puts "\nReason: policy file supplied is not a file"
+      exit 1
+    end
+    options["policy_file"] = SSHScan::Policy.from_file(options["policy"])
```
Comment (claudijd) [inline bin/ssh_scan]: This is a behavior change in which a user will always be required to specify a policy file.  Is this truly the behavior we're shooting for?

File: lib/ssh_scan/policy.rb [hunk 12]
```diff
@@ -25,5 +27,23 @@ def self.from_string(string)
...
+      Dir.glob(policies_dir + "/**/*.yml") do |policy_file|
+        next if policy_file == fallback_file
+        file = YAML.load_file(policy_file)
+        next unless file["rules"]
+        logger.info("Matching: #{policy_file}")
+        if SSHScan::RuleEngine.matchRuleset(file["rules"], file["ruleset"], result)
+          logger.info("Matched: #{policy_file}")
+          return policy_file
+        end
+      end
```
Comment (claudijd) [inline lib/ssh_scan/policy.rb]: Maybe we should log that we ended up using the fall back?

File: lib/ssh_scan/rule_engine.rb [hunk 13]
```diff
@@ -0,0 +1,94 @@
+module SSHScan
+  class RuleEngine
+    # Simple one rule matching.
+    def self.matchSingleRule(key, value, result)
```
Comment (claudijd) [inline lib/ssh_scan/rule_engine.rb]: Maybe we should start using YARD documentation for these interfaces, so it's more clear what each parameter class type is?

File: lib/ssh_scan/rule_engine.rb [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+      answer = true
```
Comment (claudijd) [inline lib/ssh_scan/rule_engine.rb]: Should the default answer be false and not true?

I'd rather fall back than incorrectly match.

File: lib/ssh_scan/rule_engine.rb [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+          when "gte" then (rvalue >= value)
+          when "lte" then (rvalue <= value)
+          else true
+        end
+      end
+      answer
+    end
+
+    # Match a simple ruleset condition using the stack based parser.
+    def self.matchRuleset(rules, ruleset, result)
```
Comment (claudijd) [inline lib/ssh_scan/rule_engine.rb]: Again, here is a good place where YARD docs would help

File: lib/ssh_scan/rule_engine.rb [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+      mem = {}
+      rules.each do |rule_name, values|
+        mem[rule_name] = matchSingleRule(values[0], values[1], result)
+      end
+      return parseBooleanRule(ruleset, mem)
+    end
+
+    # Check if o is boolean or not. Apparently is_a?(Boolean) is not a thing.
```
Comment (claudijd) [inline lib/ssh_scan/rule_engine.rb]: yeah, in Ruby, you've got True class and False classes.

File: lib/ssh_scan/rule_engine.rb [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        stack << new_tokenval
+      elsif (stack[-2] == '||') && checkBool(stack[-3]) && checkBool(stack[-1])
+        new_tokenval = stack[-1] || stack[-3]
+        3.times { stack.pop }
+        stack << new_tokenval
+      else
+        return false
+      end
+      return true
+    end
```
Comment (claudijd) [inline lib/ssh_scan/rule_engine.rb]: I'm not sure what this is doing. Maybe some more documentation around this functionality would help.

File: spec/ssh_scan/rule_engine_spec.rb [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    it "neq should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_neq", 1.99, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_neq", 2.0, result)).to be false
+    end
+
+    it "gt should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 1.3, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 2.0, result)).to be false
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Recommend using the following cases:

expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 1.99, result)).to be true
expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 2.0, result)).to be false
expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 2.1, result)).to be false

Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Recommending 1.99 because it's common and a closer boundary to check.

Recommending 2.1 because it tests the other side of the boundary.

File: spec/ssh_scan/rule_engine_spec.rb [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+    end
+
+    it "gte should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gte", 1.3, result)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Recommend replacing 1.3 with...

expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 1.99, result)).to be true

File: spec/ssh_scan/rule_engine_spec.rb [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lt", 2.1, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lt", 2.0, result)).to be false
+    end
+
+    it "lte should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 2.1, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 2.0, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 1.3, result)).to be false
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Recommend replacing...

expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 1.99, result)).to be false

File: spec/ssh_scan/rule_engine_spec.rb [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+    end
+
+    it "should ignore attributes not in results" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_lib_eq", "openssh", result)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: I would expect this to fail, as the default behavior when the key requirement is not present should be to return false, right?

  expect(SSHScan::RuleEngine.matchSingleRule("ssh_lib_eq", "openssh", result)).to be false

File: spec/ssh_scan/rule_engine_spec.rb [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+    it "should work for NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x", "x" => true)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x", "x" => false)).to be true
+    end
+
+    it "should work for AND expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => true, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => true, "y" => false)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => true)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => false)).to be false
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: I like that you're going through all the cases here, having complete coverage can give you a lot more confidence in the code not having logic/boundary bugs.

File: spec/ssh_scan/rule_engine_spec.rb [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+    it "should work for OR expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => true, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => true, "y" => false)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => false, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => false, "y" => false)).to be false
+    end
+
+    it "should work for compounded non bracketed AND, OR, NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x && y", "x" => false, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || !y", "x" => false, "y" => false)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Would recommend expanding the test cases here to include....
 
    expect(SSHScan::RuleEngine.parseBooleanRule("x || !y", "x" => false, "y" => false)).to be true
    expect(SSHScan::RuleEngine.parseBooleanRule("!x || y", "x" => false, "y" => false)).to be true
    expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => false, "y" => false)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("!x || !y", "x" => false, "y" => false)).to be true

    expect(SSHScan::RuleEngine.parseBooleanRule("x || !y", "x" => false, "y" => true)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("!x || y", "x" => false, "y" => true)).to be true
    expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => false, "y" => true)).to be true
    expect(SSHScan::RuleEngine.parseBooleanRule("!x || !y", "x" => false, "y" => true)).to be true

    expect(SSHScan::RuleEngine.parseBooleanRule("x && !y", "x" => false, "y" => false)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("!x && y", "x" => false, "y" => false)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => false)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("!x && !y", "x" => false, "y" => false)).to be true

    expect(SSHScan::RuleEngine.parseBooleanRule("x && !y", "x" => false, "y" => true)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("!x && y", "x" => false, "y" => true)).to be true
    expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => true)).to be false
    expect(SSHScan::RuleEngine.parseBooleanRule("!x && !y", "x" => false, "y" => true)).to be false

File: spec/ssh_scan/rule_engine_spec.rb [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+    end
+
+    it "should work for compounded bracketed AND, OR, NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && (y || z)",
+        "x" => false, "y" => true, "z" => false)).to be false
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Please add...

    expect(SSHScan::RuleEngine.parseBooleanRule("x && (y || z)",  "x" => false, "y" => true, "z" => true)).to be true

File: spec/ssh_scan/rule_engine_spec.rb [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+      result2 = {:ssh_version => 1.99, :ssh_lib => "dropbear"}
+      result3 = {:ssh_version => 2.1, :ssh_lib => "dropbear"}
+      rules = {
+        "a" => ["ssh_version_gt", 1.99],
+        "b" => ["ssh_lib_eq", "openssh"],
+        "c" => ["ssh_lib_eq", "dropbear"],
+        "d" => ["ssh_version_lt", 1.99]
+      }
+      ruleset = "(a && b) || (c && !d)" # !d means ssh_version_gte 1.99
+      expect(SSHScan::RuleEngine.matchRuleset(rules, ruleset, result1)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Walking through the logic out loud...

- ssh version 2.0 is gt than 1.99 (OK)
- ssh lib openssh is eq openssh (OK)

Overall: OK

File: spec/ssh_scan/rule_engine_spec.rb [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+      expect(SSHScan::RuleEngine.matchRuleset(rules, ruleset, result2)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Walking through the logic out loud...

    ssh version 1.99 is NOT gt than 1.99 (OK)
    ssh lib dropbear is eq dropbear (OK)
    ssh version 1.99 is gte than 1.99 (OK)

Overall: OK

File: spec/ssh_scan/rule_engine_spec.rb [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+      expect(SSHScan::RuleEngine.matchRuleset(rules, ruleset, result3)).to be true
```
Comment (claudijd) [inline spec/ssh_scan/rule_engine_spec.rb]: Walking through the logic out loud...

    ssh version 2.1 is gt than 1.99 (OK)
    ssh lib dropbear is NOT eq openssh (OK)
    ssh lib dropbear is eq dropbear (OK)
    ssh version 2.1 is gte than 1.99 (OK)

Overall: OK

**Comment by claudijd** [comment]
Overall, looks pretty good.  I added some suggestions about additional/modifications for test cases.  I also think that we should reconsider how we treat matches when the element is not present in the result.  The boolean/stack ranking stuff is a little hard to follow as a ruby person, but you've a good foundation for testing that leads me to believe it works as expected.

File: bin/ssh_scan (hunk 11 above)
Comment (agaurav77) [inline bin/ssh_scan]: From what I understand, the user doesn't need to always specify a policy file. But when he/she doesn't, the policy file is found on a first match basis, failing which we use the fallback file.

File: lib/ssh_scan/rule_engine.rb (hunk 13 above)
Comment (agaurav77) [inline lib/ssh_scan/rule_engine.rb]: This definitely looks promising. Opened up #319 for this. I think we should pursue using YARD in a different PR.

File: lib/ssh_scan/rule_engine.rb (hunk 14 above)
Comment (agaurav77) [inline lib/ssh_scan/rule_engine.rb]: I agree. I was looking from an AND perspective, but it is certainly possible to do this. Will change.

File: lib/ssh_scan/rule_engine.rb (hunk 17 above)
Comment (agaurav77) [inline lib/ssh_scan/rule_engine.rb]: Added some examples, comments.

File: spec/ssh_scan/rule_engine_spec.rb (hunk 21 above)
Comment (agaurav77) [inline spec/ssh_scan/rule_engine_spec.rb]: Yeah, according to default behavior as you suggested above, it should. Changing...

File: spec/ssh_scan/rule_engine_spec.rb (hunk 24 above)
Comment (agaurav77) [inline spec/ssh_scan/rule_engine_spec.rb]: Shouldn't this be false?

**Comment by agaurav77** [comment]
@claudijd should be good now, please take a look!

**Comment by claudijd** [comment]
Blocked by #307
